### PR TITLE
[presto-cli] Cancel query & exit on SIGTERM

### PR DESCRIPTION
**Pull request by nezihyigitbasi**
Currently presto cli cancels a query when it receives a `SIGINT` signal. However, when it receives a `SIGTERM` signal the cli process exits without canceling a query, leaving the query running in the cluster until it gets abandoned. This PR adds support for gracefully handling the `SIGTERM` signal in the cli.

### CODE REVIEW

File: presto-cli/src/main/java/com/facebook/presto/cli/Query.java [hunk 1]
```diff
@@ -73,19 +74,27 @@ public Query(StatementClient client)
...
-        Thread clientThread = Thread.currentThread();
-        SignalHandler oldHandler = Signal.handle(SIGINT, signal -> {
+        SignalHandler intHandler = signal -> {
             if (ignoreUserInterrupt.get() || client.isClosed()) {
                 return;
             }
             client.close();
-            clientThread.interrupt();
-        });
+            Thread.currentThread().interrupt();
```
Comment (cberner) [inline presto-cli/src/main/java/com/facebook/presto/cli/Query.java]: Is this going to do the right thing? I would guess that the handler will be invoked by a different thread than the client thread

Comment (nezihyigitbasi) [inline presto-cli/src/main/java/com/facebook/presto/cli/Query.java]: you are right. I will update the PR.

**Comment by nezihyigitbasi** [comment]
@cberner updated.

**Comment by electrum** [comment]
I don't think we need to handle `SIGTERM` explicitly. Instead, we should be able to install a shutdown hook that interrupts the main thread, which should cancel the query (the Presto JDBC driver handles interruption). We might need handling in other places like `Console.runConsole()` to exit on interrupt.

Try this in `Console` before creating `QueryRunner`:
```java
@SuppressWarnings("ClassExplicitlyExtendsThread")
private static void interruptThreadOnExit()
{
    Thread thread = Thread.currentThread();
    Runtime.getRuntime().addShutdownHook(new Thread()
    {
        @Override
        public void run()
        {
            thread.interrupt();
        }
    });
}

**Comment by nezihyigitbasi** [comment]
Shutdown hook also registers handlers for sigint and maybe some other signals, wouldn't that interfere with the sigint handler installed here? I just wanted to be specific that's why i only registered a handler for sigterm. Will give this a try.

File: presto-cli/src/main/java/com/facebook/presto/cli/Query.java [hunk 2]
```diff
@@ -84,6 +84,14 @@ public void renderOutput(PrintStream out, OutputFormat outputFormat, boolean int
         try {
             renderQueryOutput(out, outputFormat, interactive);
         }
+        catch (QueryAbortedException | RuntimeException e) {
```
Comment (nezihyigitbasi) [inline presto-cli/src/main/java/com/facebook/presto/cli/Query.java]: `StatementClient` throws a `RuntimeException` on interruption.

**Comment by nezihyigitbasi** [comment]
@electrum updated the PR to use a shutdown hook.

File: presto-client/src/main/java/com/facebook/presto/client/StatementClient.java [hunk 3]
```diff
@@ -223,7 +223,7 @@ public boolean advance()
                 }
                 catch (InterruptedException e) {
                     close();
-                    Thread.currentThread().isInterrupted();
+                    Thread.currentThread().interrupt();
```
Comment (electrum) [inline presto-client/src/main/java/com/facebook/presto/client/StatementClient.java]: This is a bug fix that should be done in a separate commit. Nice catch!

File: presto-cli/src/main/java/com/facebook/presto/cli/Console.java [hunk 4]
```diff
@@ -80,6 +81,26 @@
...
+    {
+        Thread thread = Thread.currentThread();
+        Runtime.getRuntime().addShutdownHook(new Thread()
+        {
+            @Override
+            public void run()
+            {
+                thread.interrupt();
+                try {
+                    thread.join(MAIN_JOIN_TIMEOUT);
```
Comment (electrum) [inline presto-cli/src/main/java/com/facebook/presto/cli/Console.java]: Why is this needed?

Comment (nezihyigitbasi) [inline presto-cli/src/main/java/com/facebook/presto/cli/Console.java]: because the thread running the shutdown hook will just call `interrupt()` (and just set a flag) and the vm may exit even before the main thread checks for the interrupted status, which I have seen during testing.

**Comment by nezihyigitbasi** [comment]
@electrum addressed @martint 's comments.

**Comment by electrum** [comment]
I cherry picked the interrupt status fix. Will review the rest soon.

**Comment by nezihyigitbasi** [comment]
@electrum Just rebased and made minor changes. Can you please review?

**Comment by nezihyigitbasi** [comment]
@electrum rebased & simplified.

**Comment by electrum** [comment]
I was testing this out and noticed that without this, the CLI exits immediately when killed and properly resets ctrl-C to be the interrupt character. With this change, the CLI prints "Query aborted by user", prints the "presto>" prompt, then exits, but doesn't restore the interrupt character.

I need to figure out exactly what is happening in both scenarios. Any ideas?

**Comment by nezihyigitbasi** [comment]
@electrum jline seems to be executing `stty` commands under the hood, see `TerminalLineSettings::stty()`. I need to debug to understand how to recover it though.

**Comment by nezihyigitbasi** [comment]
I think what's happening is jline tries to recover the interrupt character and creates a process to run the command below (this command recovers the intr character, you can confirm it by running that command when the interrupt character is disabled -- you can check the result with `stty -a`)

```
stty gfmt1:cflag=4b00:iflag=6b02:lflag=200005cb:oflag=3:discard=f:dsusp=19:eof=4:eol=ff:eol2=ff:erase=7f:intr=3:kill=15:lnext=16:min=1:quit=1c:reprint=12:start=11:status=14:stop=13:susp=1a:time=0:werase=17:ispeed=9600:ospeed=9600
```

but the shutdown hook doesn't wait for this process to finish.

**Comment by nezihyigitbasi** [comment]
@electrum how do you feel about enabling the interrupt character in the shutdown hook? We can refactor a bit to pass the `Terminal` instance (that we can get from `LineReader`) to the shutdown hook and call:

```
((UnixTerminal) terminal).enableInterruptCharacter();
```

Or do you have another suggestion?

**Comment by electrum** [comment]
Merged, thanks!

**Comment by electrum** [comment]
After lots and lots of debugging, I finally figured out what was happening: when the JVM gets a SIGTERM, it runs the shutdown hooks, then exits (even if there are non-daemon threads still running).

The thread join would wait the full timeout and return, with the main thread still running (i.e. the join times out). The main thread would be sitting at the readline prompt when the JVM exits, thus never running the cleanup code.

I fixed that by changing the readline loop to watch for a flag that gets set by the shutdown hook. This allows the main thread to exit normally and thus perform all the cleanup. It also makes the thread join quick because the main thread actually exits.

**Comment by nezihyigitbasi** [comment]
@electrum thanks!
